src/engine/profiler.cc

```diff
@@ -45,12 +50,11 @@ Profiler::Profiler()
   : state_(kNotRunning), enable_output_(false), filename_("profile.json") {
   this->init_time_ = NowInUsec();
 
-  // TODO(ziheng) get device number during execution
-  int kMaxNumCpus = 64;
-  this->cpu_num_ = kMaxNumCpus;
+  this->cpu_num_ = std::thread::hardware_concurrency();
 #if MXNET_USE_CUDA
-  int kMaxNumGpus = 32;
-  this->gpu_num_ = kMaxNumGpus;
+  int gpu_num = 0;
+  CUDA_CALL(cudaGetDeviceCount(&gpu_num));
```

Comment:
We intentionally used a constant number here because we don't want to initialize cuda context before gpu is used for the first time

Reply:
@piiswrong Thank you for your comment.
I understood the purpose of using the constant number.
How about dynamically setting only the number of CPUs?
More than 64 CPUs are not ordinary but not impossible.